### DISCUSSION TITLE
Global illumination

### DISCUSSION THREAD

**Discussion by apan65** [Try/OpenGothic#373] [General]
Так как рендеринг на вулкане, то волну интереса к этой версии движка могут поднять лучи. Мне и самому интересно как будет выглядеть игра с полным рейтрейсингом.

**Comment by Try** [comment]
> Мне и самому интересно как будет выглядеть игра с полным рейтрейсингом.

Em, exactly same as is:
https://github.com/Try/OpenGothic/issues/261
https://www.youtube.com/watch?v=TpayMkyZ58Y

We do not support RT+animation, since neither DX or Vulkan can do it straight.

**Comment by apan65** [comment]
> > Мне и самому интересно как будет выглядеть игра с полным рейтрейсингом.
> 
> Em, exactly same as is: #261 https://www.youtube.com/watch?v=TpayMkyZ58Y
> 
> We do not support RT+animation, since neither DX or Vulkan can do it straight.

I mean global illumination, not shadows.

**Comment by Try** [comment]
> I mean global illumination, not shadows.

That a quite orthogonal problem :D and very big one. There no easy way to implement a proper GI over weekend.
So far I've been looking toward to:
1. http://mrakobes.com/Nikolay.Stefanov.GDC.2016.pdf
   - (+) Probe based
   - (+) Probes can be placed automatically
   - (-) Doesn't work as-is in realtime (The Division uses ~1.07GB of cache)
2. https://www.youtube.com/watch?v=h1ocYFrtsM4
   - (+) Surfels based
   - (+) Like Frostbite, OpenGothic has global matrix ssbo (5:12 in video)
   - (-) Doesn't work as-is without rayQuery/rayPipeline

There is also new approach from AMD: https://gpuopen.com/download/publications/GPUOpen2022_GI1_0.pdf
Haven't look too deep, seems to be yet another surfels-based approach.

**Comment by apan65** [comment]
i don't know anything about coding but is it helps? https://github.com/sultim-t/RayTracedGL1

**Comment by Try** [comment]
Nope. While it's nice project, that wont help at all

**Comment by apan65** [comment]
> Nope. While it's nice project, that wont help at all
RTXDI, ReSTIR GI, RTXGI, or anything from this?
https://github.com/orgs/NVIDIAGameWorks

**Comment by Try** [comment]
In my opinion ReSTIR as-is is just bad chose - surfels(Frostbite) are just better overall.

I've give more thoughts on GI in general, we probably do not have pre-requirements to even start with that:
All GI solutions assume PBR pipeline. We have natural limitations with PBR:
* No roughness/metallic/albedo textures. Workaround: roughness=1, metallic=0, albedo=diffuse
* Must go linear. At this point rendering works, same as DX7, in gamma color space. Gamma give look similar to how it was in vanilla, at cost of correctness. 
* Linear color space doesn't look as vanilla. Here is [UE5 example](https://www.reddit.com/r/unrealengine/comments/u55ykx/gothic_2_in_unreal_engine_5_day_7_ue_newbie/). Generally coloring is similar to results in my tests - game become 'orangish' 
* Must have HDR(+tonemapping). HDR makes AA implementation even more painful than it is.
* Probably not everyone will be happy to have HDR. Supporting HDR and non-HDR in a same time is difficult
* Lights have arbitrary luminosity. This is content issue of vanilla, so need some solution to convert that into Lux/Lumens
* Same for particles(fire) - they too 'dark'

While this a challenging and interesting stuff to solve, it's not gonna be fast. Or easy :D

PS:
There is somewhat easy 80% solution, can be irradiance map - it can be generated from SkyLUT, and give quite an impact on lighting.

**Comment by Try** [comment]
Testing linear rendering today. Main idea is to move toward correct lighting step-by-step, where GI is very last one.
Not everything works well out of the box, in any case here some screen-shoots to illustrate.
![изображение](https://user-images.githubusercontent.com/122894/209574279-4393cb60-11cb-454b-84bd-b6b1b05381ab.png)

Had to undo quite a few old hacks, related to gamma. Image still... well can be better:
On left fire is almost invisible - tested with and without gamma 
On a right - shafts are better; but landscape feels too dark.

Another sky-case:
![изображение](https://user-images.githubusercontent.com/122894/209574921-783b46fa-61d4-4a9f-91ea-fab4c4273f5b.png)
Old colony just happen to be good case it seems.

Farmland(6:52am): 
![изображение](https://user-images.githubusercontent.com/122894/209575386-de9def3e-6d78-4d6d-bbf1-e0418de0f2f4.png)

Overall: ambient light looks too dark

**Reply by apan65** [reply]
Your volumetric lightening so good, this is what dx11 render extension strongly needed
